Docs for CLP Pointer permissions.

BufferUnderflower opened the pull request:
parse-community/parse-server#6351
parse-community/parse-server#6352

BufferUnderflower committed:
[WIP] Class Level Permissions

parse-community/parse-server#6351
parse-community/parse-server#6352

BufferUnderflower committed:
updated

BufferUnderflower committed:
updates common security section

BufferUnderflower committed:
revert bundle.js

BufferUnderflower committed:
bundle.js

BufferUnderflower committed:
bundlejs

BufferUnderflower committed:
reset

BufferUnderflower committed:
bunfle keeps resetting idk

BufferUnderflower committed:
Update bundle.js

BufferUnderflower committed:
Update security.md

BufferUnderflower committed:
Update class-level-permissions.md

TomWFox commented:
Wow, this is quite substantial, reviewing this is on my to-do list now that the parse server pr has been merged but this won’t be merged until the next release.

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+Allowed entries for operations are:
+
+* `*`  -  [Public](#public-access)
```
```suggestion
* `*` - [Public](#public-access)
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+### Public access
+
+`*` - Allows anyone despite authentication status to execute operation.
```
```suggestion
`*` - Allows anyone regardless of authentication status to execute the operation.
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+### Users, Roles
+
+This works exactly as ACL's
```
```suggestion
This works exactly like ACLs.
```

TomWFox commented on _includes/common/security.md:
````diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+}
+```
+
+### Requires Authentication permission (requires parse-server  >= 2.3.0)
````
```suggestion
### Requires Authentication permission
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+### Requires Authentication permission (requires parse-server  >= 2.3.0)
+
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
```
```suggestion
Starting with version 2.3.0, Parse Server introduced a new Class Level Permission `requiresAuthentication`. It prevents any non authenticated user from performing the action protected by the CLP.
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+### Requires Authentication permission (requires parse-server  >= 2.3.0)
+
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
+This CLP prevents any non authenticated user from performing the action protected by the CLP.
```
Moved to the paragraph above...
```suggestion
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
+This CLP prevents any non authenticated user from performing the action protected by the CLP.
+
+If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` class level permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the CLP:
```
First sentence seemed to be largely repeating the earlier explanation...
```suggestion
For example, if you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you could set the following CLP:
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+Effects:
+
+* Non authenticated users won't be able to do anything.
+* Authenticated users (any user with a valid sessionToken) will be able to read all the objects in that class
```
```suggestion
* Authenticated users (any user with a valid `sessionToken`) will be able to read all the objects in that class.
```

TomWFox reviewed (changes requested):
Sorry for the delay, here's what I've got so far. Also is this still a work in progress?

For my reference I got down to line 162.

TomWFox commented:
@BufferUnderflower If you've got maintainer commits turned on I could commit these suggestions and others myself. Otherwise could you look over my review? - there is going to be a Parse Server release soon so I'd like to get this ready to go.

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+## Pointer Permissions
+
+While permissions discussed before let you explicitly target a user by id or a role, Pointer Permissions let you dynamically enforce permissions without knowing the id or assigning roles in advance. Instead you define one or more field names of this class, and any users pointed by such fields of a particular object are granted the permission.
```
```suggestion
While the permissions discussed before let you explicitly target a user by `id` or a role, Pointer Permissions let you dynamically enforce permissions without knowing the `id` or assigning roles in advance. Instead you define one or more field names of a class, and any users pointed to by such fields of a particular object are granted the permission.
```

BufferUnderflower replied:
Well there is a field that points (holds pointers to) user(s) => users are pointed by (or may be 'in') these fields. What do you think?  `to` seems a bit misleading to me

TomWFox replied:
`to` seems right to me but perhaps I'm misunderstanding how it works

BufferUnderflower replied:
Imagine u have a PullRequest class, with one of its columns 'reviewers'.  in 'reviewers' there are ids of various people having rights to review it. Given a `Pointer<_User>` (_your_id_) is listed in `reviewers` column. Is it ok to say you are pointed to this column? Or you are pointed by/in this column?

TomWFox replied:
Ok, I've got it. Personally I think 'pointed to by this column' makes more sense.

TomWFox replied:
Changes applied in commit 0bb548e

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,428 @@ You can configure the client's ability to perform each of the following operatio
+
+While permissions discussed before let you explicitly target a user by id or a role, Pointer Permissions let you dynamically enforce permissions without knowing the id or assigning roles in advance. Instead you define one or more field names of this class, and any users pointed by such fields of a particular object are granted the permission.
+
+Pointer permissions are a special type of class-level permission that create a virtual ACL on every object in a class, based on users stored in pointer fields on those objects. For example, given a class with an `owner` field, setting a read pointer permission on `owner` will make each object in the class only readable by the user in that object's `owner` field. For a class with a `sender` and a  ...
```
```suggestion
Pointer Permissions are a special type of class-level permission that create a virtual ACL on every object in a class, based on the users stored in pointer fields on those objects. For example, given a class with an `owner` field, setting a read pointer permission on `owner` will make each object in the class only readable by the user in that object's `owner` field. For a class with a `sender` and a `reciever` field, a read pointer permission on the `receiver` field and a read and write pointer permission on the `sender` field will make each object in the class readable by the user in the `sender` and `receiver` field, and writable only by the user in the `sender` field.
```

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update security.md

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+In case of `Array`, only items that are of `Pointer<_User>` type are evaluated, other items silently ignored.
+
+You can think of it as a virtual ACL or a dynamic role defined per-object in its own field.
```
Seems repetitive...
```suggestion
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+You can think of it as a virtual ACL or a dynamic role defined per-object in its own field.
+
+There are two ways you can set Pointer Permission in schema:
```
```suggestion
There are two ways you can set Pointer Permissions in schema:
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+There are two ways you can set Pointer Permission in schema:
+
+* [Using granular permissions](#granular-pointer-permissions) - `pointerFields` *requires Parse-Server v3.11 and above*
```
```suggestion
* [Using granular permissions](#granular-pointer-permissions) - `pointerFields` *requires Parse Server v3.11 or above*
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+* [Using granular permissions](#granular-pointer-permissions) - `pointerFields` *requires Parse-Server v3.11 and above*
+* [Using grouped permissions](#grouped-pointer-permissions) - `readUserFields`/`writeUserFields`
+
+⚠️ `create` operation can't be allowed by pointer permissions, because there is literally no object to check it's field untill it is created;
```
```suggestion
⚠️ The `create` operation can't be controlled by Pointer Permissions, because there is no object to check the field of until it is created;
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+⚠️ `create` operation can't be allowed by pointer permissions, because there is literally no object to check it's field untill it is created;
+
+⚠️ `addField`  grants permission to only update an object with a new field, but it is advised to set addField permission using other means (e.g. restrict to a role or particular admin user by id).
```
```suggestion
⚠️ The `addField` operation  grants permission to only update an object with a new field, but it is advised to set `addField` permissions using other means (e.g. restrict to a role or particular admin user by `id`).
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+* anyone is allowed to `create` objects.
+* `feedA` can be viewed (`get`,`find`) only by **Alice**. (pointed in owners field)
+* `feedB` can be viewed (`get`,`find`) both by **Bob** and **Alice**. (Bob is pointed in `owners`. Alice is pointed in `subscribers`)
```
```suggestion
* `feedB` can be viewed (`get`,`find`) both by **Bob** and **Alice** (Bob is pointed to in `owners`, Alice is pointed to in `subscribers`).
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+* anyone is allowed to `create` objects.
+* `feedA` can be viewed (`get`,`find`) only by **Alice**. (pointed in owners field)
+* `feedB` can be viewed (`get`,`find`) both by **Bob** and **Alice**. (Bob is pointed in `owners`. Alice is pointed in `subscribers`)
+* only owners are allowed to `update` and `delete`.
```
```suggestion
* Only owners are allowed to `update` and `delete`.
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+In the example above:
+
+* anyone is allowed to `create` objects.
+* `feedA` can be viewed (`get`,`find`) only by **Alice**. (pointed in owners field)
```
```suggestion
* `feedA` can be viewed (`get`,`find`) only by **Alice** (pointed to in `owners` field).
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,427 @@ You can configure the client's ability to perform each of the following operatio
+
+In the example above:
+
+* anyone is allowed to `create` objects.
```
```suggestion
* Anyone is allowed to `create` objects.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,115 @@
+
+### Requires Authentication permission
+
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
```
```suggestion
Starting version 2.3.0, Parse Server introduces a new Class Level Permission called `requiresAuthentication`. This CLP prevents any non authenticated user from performing the action protected by the CLP.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,115 @@
+### Requires Authentication permission
+
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
+This CLP prevents any non authenticated user from performing the action protected by the CLP.
```
```suggestion
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,115 @@
+Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.
+This CLP prevents any non authenticated user from performing the action protected by the CLP.
+
+If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` class level permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the CLP:
```
```suggestion
If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` Class Level Permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the following CLP:
```

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/parse-server/class-level-permissions.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/parse-server/class-level-permissions.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/parse-server/class-level-permissions.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Apply suggestions from code review

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
Update _includes/common/security.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,114 @@
+
+### Public access
+
+`*` - Allows anyone despite authentication status to execute operation.
```
```suggestion
`*` - Allows anyone regardless of authentication status to execute the operation.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,114 @@
+
+### Users, Roles
+
+This works exactly as ACL's
```
```suggestion
This works exactly like ACL's.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -1,23 +1,114 @@
+
+Starting version 2.3.0, Parse Server introduces a new Class Level Permission called `requiresAuthentication`. This CLP prevents any non authenticated user from performing the action protected by the CLP.
+
+If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` Class Level Permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the following CLP:
```
```suggestion
Starting with version 2.3.0, Parse Server introduced a new Class Level Permission `requiresAuthentication`. It prevents any non authenticated user from performing the action protected by the CLP.

For example, if you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you could set the following CLP:
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -28,4 +119,163 @@ If you want to restrict access to a full class to only authenticated users, you
+* Authenticated users (any user with a valid sessionToken) will be able to read all the objects in that class
+* Users belonging to the admin role, will be able to perform all operations.
+
+⚠️ Note that this is in no way securing your content, if you allow anyone to log in to your server, any client will still be able to query this object.
```
```suggestion
Effects:

* Non authenticated users won't be able to do anything.
* Authenticated users (any user with a valid `sessionToken`) will be able to read all the objects in that class.
* Users belonging to the admin role, will be able to perform all operations.

⚠️ Note that this is in no way securing your content, if you allow anyone to login to your server, every client will still be able to query this object.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -28,4 +119,163 @@ If you want to restrict access to a full class to only authenticated users, you
+
+## Pointer Permissions
+
+While permissions discussed before let you explicitly target a user by id or a role,
+Pointer Permissions let you dynamically enforce permissions without knowing the id or assigning roles in advance. Instead you define one or more field names of this class, and any users pointed by such fields of a particular object are granted the permission.
+
+To use this feature, a field must:
+
+* already exist in this collection's schema
+* be of either `Pointer<_User>` or `Array` type
+
+In case of `Array`, only items that are of `Pointer<_User>` type are evaluated, other items silently ignored.
+
+You can think of it as a virtual ACL or a dynamic role defined per-object in its own field.
+
+There are two ways you can set Pointer Permission in schema:
+
+* [Using granular permissions](#granular-pointer-permissions) - `pointerFields` *requires Parse-Server v3.11 and above*
+* [Using grouped permissions](#grouped-pointer-permissions) - `readUserFields`/`writeUserFields`
+
+⚠️ `create` operation can't be allowed by pointer permissions, because there is literally no object to check it's field untill it is created;
+
+⚠️ `addField`  grants permission to only update an object with a new field, but it is advised to set addField permission using other means (e.g. restrict to a role or particular admin user by id).
```
There is quite a discrepancy between here and the corresponding common section...
https://github.com/parse-community/docs/pull/700/files#diff-22c65bfd6614d949ec20e9d4461ce3cfR163-R169

This part needs some of the grammatical improvements that I suggested in the common section although I wouldn't add all the explanation - in fact I would suggest removing some of the long example paragraphs from the common doc on this section as they seem a bit repetitive / overly wordy.

BufferUnderflower replied:
This brought me to a question - why do we have CLP section in Parse-Server guide in the first place?  It seems to have info about server as a binary, its setup and launch, while CLP is more about schema level feature. I suggest to remove CLP from server guide as a whole.

TomWFox replied:
good question, I think the Parse Server guide has been under loved so only contains limited info about basic / required setup - and there are discrepancies in what is included in the Parse Server README and the guide - I did aim to consolidate all the info into the guide but didn't get around to it.

I'm wondering whether we could remove the CLP section in the server guide in favour of adding the common security doc but I'm not sure how this would work in practice with the code sample etc?

BufferUnderflower replied:
In fact the code samples are mostly schema configs - plain JSON objects. And a few examples come with several demo setup lines (that i've already translated to most sdk languages in this pr). 
Anyway clp and protected fields are used by database admin rather than client code or any of SDKs. I also think it could be very helpful and time saving if we maintain such examples as a shared postman collection.

TomWFox replied:
I see that.

I think adding the common security doc to the server guide would be good, that way there is only one 'copy' to be maintained and its available on all the guides which would encourage devs to think about security when they are implementing their client side.

TomWFox replied:
So are you happy with that or do you have other ideas?

I think it needs to be in there server guide in some way as it is related to setup - especially as you have gone to all the effort of writing it and I don't think it would make sense to have it in the client guides but not the server one?

BufferUnderflower replied:
I agree we should merge it and only have a single copy as you proposed. As for Parse-Server Guide probably REST-flavored examples should fit best.

TomWFox replied:
Cool, REST example would make sense. I'll leave that with you then, or if you've got maintainer commits turned on I could take a look later.

BufferUnderflower replied:
I believe i created this pr with edits enabled. Not sure if there is anything else i should allow?
<img width="261" alt="image" src="https://user-images.githubusercontent.com/2188841/75628552-3c941000-5beb-11ea-9d89-d98ef63d0181.png">
 I'm planning to work on it later today anyway

TomWFox replied:
Yes that’s the one, the annoying thing is that on GitHub there is no indication to me whether you have that ticked or not.

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -28,4 +119,163 @@ If you want to restrict access to a full class to only authenticated users, you
+* anyone is allowed to `create` objects.
+* `feed1` can be viewed (`get`,`find`) only by **Alice**. (pointed in owners field)
+* `feed2` can be viewed (`get`,`find`) both by **Bob** and **Alice**. (Bob is pointed in `owners`. Alice is pointed in `subscribers`)
+* only owners are allowed to `update` and `delete`.
```
```suggestion
In the example above:

* Anyone is allowed to `create` objects.
* `feedA` can be viewed (`get`,`find`) only by **Alice** (pointed to in `owners` field).
* `feedB` can be viewed (`get`,`find`) both by **Bob** and **Alice** (Bob is pointed to in `owners`, Alice is pointed to in `subscribers`).
* Only owners are allowed to `update` and `delete`.
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -28,4 +119,163 @@ If you want to restrict access to a full class to only authenticated users, you
+* `delete`,
+* `addField`
+
+Same scheme as for previous example can be defined shorter:
```
```suggestion
**`readUserFields`**:

* `get`
* `find`
* `count`

**`writeUserFields`**:

* `update`
* `delete`
* `addField`

The same scheme as for previous example can be defined in a more compact manner:
```

TomWFox commented on _includes/common/security.md:
```diff
@@ -58,6 +58,426 @@ You can configure the client's ability to perform each of the following operatio
+* `delete`,
+* `addField`
+
+Same scheme as for previous example can be defined shorter:
```
```suggestion
**`readUserFields`**:

* `get`
* `find`
* `count`

**`writeUserFields`**:

* `update`
* `delete`
* `addField`

The same scheme as for previous example can be defined in a more compact manner:
```

TomWFox commented on _includes/parse-server/class-level-permissions.md:
```diff
@@ -28,4 +119,163 @@ If you want to restrict access to a full class to only authenticated users, you
+
+Now lets look at example that uses Pointer Permissions. Say we have a `Post` class, with an object, `myPost`. There are 2 users in our app, `poster`, and `viewer`. Lets say we add a pointer permission that gives anyone in the `Creator` field of the `Post` class read and write access to the object, and for the `myPost` object, `poster` is the user in that field. There is also an ACL on the object  ...
+
+Because of the complex interaction between CLPs, Pointer Permissions, and ACLs, we recommend being careful when using them together. Often it can be useful to use CLPs only to disable all permissions for a certain request type, and then using Pointer Permissions or ACLs for other request types. For example, you may want to disable Delete for a `Photo` class, but then put a Pointer Permission on ` ...
```
I'll look at this properly later but at first glance this seems very dense. Would it be possible to shorten this section or replace some of it with a diagram or schema.

BufferUnderflower replied:
agree, I believe there is some room to make it more concise. May be code will read easier than words

TomWFox reviewed (changes requested):
We're getting there! I've looked over all of it now, there may be a few more changes after this round but it's looking pretty good I'd say 🙂

Multi-line suggestions are a thing now which is nice 😅

BufferUnderflower committed:
Apply suggestions from code review

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

BufferUnderflower committed:
acl-clp and  protected fields

leaving a draft for protected fields, needs some more work for sure but its at least something

BufferUnderflower committed:
Update class-level-permissions.md

BufferUnderflower committed:
Update class-level-permissions.md

BufferUnderflower committed:
Update class-level-permissions.md

BufferUnderflower commented:
A note on protected fields. It's probably too early to disclose the documentation, since it's pretty far for being production ready. I'm now working on some tests to show what's left to be done there. Main concerns: 
* We don't cover create/update operations, and they're pretty complicated to handle. 
* As of now protection can be bypassed using aggregate queries.

TomWFox commented:
Ok, thanks for that info.

Perhaps if it's not too difficult you could remove the content about protected fields from this PR and make a separate draft PR for when its more production ready or do you want to hold pack this whole PR until those issues have been addressed on parse server?

BufferUnderflower committed:
Move protected fields to separate pr

BufferUnderflower commented:
There is no need to hold CLP, in fact it would be nice to finish it asap and merge, because protected fields would require lots of love before being ready i'm afraid.

BufferUnderflower commented:
Probably it would be wise to publish some warning in docs against use for sensitive data temporarily, since it appears in 4.0 changelog and there were several issues by folks interested in it.

TomWFox commented:
If your right about the issues with protected fields I would absolutely agree that warnings should be given. Feel free to open an issue/pr on parse server for that.

TomWFox commented:
@BufferUnderflower Sorry this has been neglected, a final review of this is on my to-do list!

awgeorge commented on _includes/common/security.md:
```diff
@@ -58,6 +58,426 @@ You can configure the client's ability to perform each of the following operatio
+      "s0meUs3r1d": true, // key must be an id of `_User`
+      "role:admin": true, // key must be `role:${role_name}`
+      "requiresAuthentication": true, // any authenticated users
+      "pointerFields": ["onwer", "followers"] // pointer permissions
```
owner.
